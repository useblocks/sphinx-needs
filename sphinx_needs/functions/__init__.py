from sphinx_needs.functions.common import (
    calc_sum,
    check_linked_values,
    copy,
    echo,
    links_from_content,
    test,
)
from sphinx_needs.functions.functions import (  # noqa: F401
    FunctionParsingException,
    execute_func,
    find_and_replace_node_content,
    register_func,
    resolve_dynamic_values,
    resolve_variants_options,
)

needs_common_functions = [test, echo, copy, check_linked_values, calc_sum, links_from_content]
