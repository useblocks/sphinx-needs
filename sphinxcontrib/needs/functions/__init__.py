from sphinxcontrib.needs.functions.common import (
    calc_sum,
    check_linked_values,
    copy,
    echo,
    links_from_content,
    test,
)
from sphinxcontrib.needs.functions.functions import (  # noqa: F401
    FunctionParsingException,
    execute_func,
    find_and_replace_node_content,
    register_func,
    resolve_dynamic_values,
)

needs_common_functions = [test, echo, copy, check_linked_values, calc_sum, links_from_content]
