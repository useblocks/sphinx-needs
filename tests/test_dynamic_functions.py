import json
import os
from pathlib import Path

import pytest
from sphinx import version_info
from sphinx.util.console import strip_colors
from syrupy.filters import props

from sphinx_needs.functions.functions import (
    DynamicFunctionParsed,
    FunctionParsingException,
    NeedAttribute,
)


@pytest.mark.parametrize(
    "func_str,expected_name,expected_args,expected_kwargs",
    [
        ("my_function()", "my_function", (), ()),
        ("my_function(1, 2, 3.0)", "my_function", (1, 2, 3.0), ()),
        ('my_function("a", "b", "c")', "my_function", ("a", "b", "c"), ()),
        ("my_function(True, False)", "my_function", (True, False), ()),
        ("my_function(need.a)", "my_function", (NeedAttribute("a"),), ()),
        (
            'my_function(1, "a", True)',
            "my_function",
            (1, "a", True),
            (),
        ),
        (
            'my_function(1, "a", True, kwarg1=2, kwarg2=3.0, kwarg3="b", kwarg4=False, kwarg5=need.b)',
            "my_function",
            (1, "a", True),
            (
                ("kwarg1", 2),
                ("kwarg2", 3.0),
                ("kwarg3", "b"),
                ("kwarg4", False),
                ("kwarg5", NeedAttribute("b")),
            ),
        ),
        (
            'my_function(kwarg1=[1, 2, 3], kwarg2=[1.0, 2.0, 3.0], kwarg3=["a", "b", "c"], kwarg4=[True, False])',
            "my_function",
            (),
            (
                ("kwarg1", [1, 2, 3]),
                ("kwarg2", [1.0, 2.0, 3.0]),
                ("kwarg3", ["a", "b", "c"]),
                ("kwarg4", [True, False]),
            ),
        ),
    ],
)
def test_dynamic_function_from_string(
    func_str, expected_name, expected_args, expected_kwargs
):
    dynamic_func = DynamicFunctionParsed.from_string(func_str, allow_need=True)

    assert dynamic_func.name == expected_name
    assert dynamic_func.args == expected_args
    assert dynamic_func.kwargs == expected_kwargs


@pytest.mark.parametrize(
    "func_str,error_message",
    [
        ("", "Not a function call"),
        ("x.", "Not a function call"),
        ("xx", "Not a function call"),
        ("dunc(None)", "Unsupported arg 0 value type"),
        ("dunc([None])", "Unsupported arg 0 item 0 value type"),
        ("dunc(x)", "Unsupported arg 0 value type"),
        ("dunc(1, x.y)", "Unsupported arg 1 value type"),
        ("dunc(func())", "Unsupported arg 0 value type"),
        ("dunc(x=x)", "Unsupported kwarg 'x' value type"),
        ("dunc(1, y=x.y)", "Unsupported kwarg 'y' value type"),
        ("dunc(z=func())", "Unsupported kwarg 'z' value type"),
        ("dunc(x=None)", "Unsupported kwarg 'x' value type"),
        ("dunc(x=[None])", "Unsupported kwarg 'x' item 0 value type"),
        ("dunc([need.x])", "Unsupported arg 0 item 0 value type"),
        ("dunc(x=[1, need.x])", "Unsupported kwarg 'x' item 1 value type"),
    ],
)
def test_dynamic_function_from_string_fail(func_str, error_message):
    with pytest.raises(FunctionParsingException, match=error_message):
        DynamicFunctionParsed.from_string(func_str, allow_need=True)


@pytest.mark.parametrize(
    "test_app",
    [
        {
            "buildername": "html",
            "srcdir": "doc_test/doc_dynamic_functions",
            "no_plantuml": True,
        }
    ],
    indirect=True,
)
def test_doc_dynamic_functions(test_app, snapshot):
    app = test_app
    app.build()

    warnings = strip_colors(
        app._warning.getvalue().replace(str(app.srcdir) + os.sep, "srcdir/")
    ).splitlines()
    assert warnings == [
        'srcdir/index.rst:11: WARNING: The `need_func` role is deprecated. Replace with :ndf:`copy("id")` instead. [needs.deprecated]',
        'srcdir/index.rst:40: WARNING: The `need_func` role is deprecated. Replace with :ndf:`copy("id")` instead. [needs.deprecated]',
        'srcdir/index.rst:44: WARNING: The `need_func` role is deprecated. Replace with :ndf:`copy("id")` instead. [needs.deprecated]',
        "srcdir/index.rst:23: WARNING: Dynamic function in list field 'tags' is surrounded by text that will be omitted: \"[[copy('title')]]omitted\" [needs.dynamic_function]",
        "srcdir/index.rst:52: WARNING: Function string 'test(need.unknown)' could not be parsed: need has no attribute 'unknown' [needs.dynamic_function]",
        'srcdir/index.rst:9: WARNING: The [[copy("id")]] syntax in need content is deprecated. Replace with :ndf:`copy("id")` instead. [needs.deprecated]',
        'srcdir/index.rst:27: WARNING: The [[copy("tags")]] syntax in need content is deprecated. Replace with :ndf:`copy("tags")` instead. [needs.deprecated]',
        "srcdir/index.rst:33: WARNING: The [[copy('id')]] syntax in need content is deprecated. Replace with :ndf:`copy('id')` instead. [needs.deprecated]",
        "srcdir/index.rst:38: WARNING: The [[copy('id')]] syntax in need content is deprecated. Replace with :ndf:`copy('id')` instead. [needs.deprecated]",
        "srcdir/index.rst:44: WARNING: Error while executing function 'copy': Need not found [needs.dynamic_function]",
        "srcdir/index.rst:44: WARNING: Error while executing function 'copy': Need not found [needs.dynamic_function]",
    ]

    json_data = Path(app.outdir, "needs.json").read_text()
    needs = json.loads(json_data)
    assert needs == snapshot(exclude=props("created", "project", "creator"))


@pytest.mark.parametrize(
    "test_app",
    [{"buildername": "html", "srcdir": "doc_test/doc_df_calc_sum"}],
    indirect=True,
)
def test_doc_df_calc_sum(test_app):
    app = test_app
    app.build()
    html = Path(app.outdir, "index.html").read_text()
    assert "43210" in html  # all hours
    assert "3210" in html  # hours of linked needs
    assert "210" in html  # hours of filtered needs


@pytest.mark.parametrize(
    "test_app",
    [{"buildername": "html", "srcdir": "doc_test/doc_df_check_linked_values"}],
    indirect=True,
)
def test_doc_df_linked_values(test_app):
    app = test_app
    app.build()
    html = Path(app.outdir, "index.html").read_text()
    assert "all_good" in html
    assert "all_bad" not in html
    assert "all_awesome" in html


@pytest.mark.parametrize(
    "test_app",
    [
        {
            "buildername": "html",
            "srcdir": "doc_test/doc_df_user_functions",
            "no_plantuml": True,
        }
    ],
    indirect=True,
)
def test_doc_df_user_functions(test_app):
    app = test_app
    app.build()

    warnings = strip_colors(
        app._warning.getvalue().replace(str(app.srcdir) + os.sep, "srcdir/")
    ).splitlines()
    # print(warnings)
    expected = [
        "srcdir/index.rst:10: WARNING: Return value of function 'bad_function' is of type <class 'object'>. Allowed are str, int, float, list [needs.dynamic_function]",
        "srcdir/index.rst:8: WARNING: The [[my_own_function()]] syntax in need content is deprecated. Replace with :ndf:`my_own_function()` instead. [needs.deprecated]",
        "srcdir/index.rst:14: WARNING: The [[bad_function()]] syntax in need content is deprecated. Replace with :ndf:`bad_function()` instead. [needs.deprecated]",
        "srcdir/index.rst:14: WARNING: Return value of function 'bad_function' is of type <class 'object'>. Allowed are str, int, float, list [needs.dynamic_function]",
        "srcdir/index.rst:16: WARNING: The [[invalid]] syntax in need content is deprecated. Replace with :ndf:`invalid` instead. [needs.deprecated]",
        "srcdir/index.rst:16: WARNING: Function string 'invalid' could not be parsed: Not a function call [needs.dynamic_function]",
        "srcdir/index.rst:18: WARNING: The [[unknown()]] syntax in need content is deprecated. Replace with :ndf:`unknown()` instead. [needs.deprecated]",
        "srcdir/index.rst:18: WARNING: Unknown function 'unknown' [needs.dynamic_function]",
    ]
    if version_info >= (7, 3):
        warn = "WARNING: cannot cache unpickable configuration value: 'needs_functions' (because it contains a function, class, or module object)"
        if version_info >= (8, 0):
            warn += " [config.cache]"
        if version_info >= (8, 2):
            warn = warn.replace("unpickable", "unpickleable")
        expected.insert(0, warn)
    assert warnings == expected

    html = Path(app.outdir, "index.html").read_text()
    assert "Awesome" in html
