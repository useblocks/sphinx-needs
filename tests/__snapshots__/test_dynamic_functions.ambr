# serializer version: 1
# name: test_doc_dynamic_functions[test_app0]
  dict({
    'current_version': '',
    'versions': dict({
      '': dict({
        'needs': dict({
          'SP_TOO_001': dict({
            'content': '''
              This is id [[copy("id")]]
              
              This is also id :need_func:`[[copy("id")]]`
              
              This is the best id :ndf:`copy("id")`
            ''',
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'SP_TOO_001',
            'lineno': 4,
            'section_name': 'DYNAMIC FUNCTIONS',
            'sections': list([
              'DYNAMIC FUNCTIONS',
            ]),
            'status': 'implemented',
            'tags': list([
              'test',
              'test2',
            ]),
            'title': 'Command line interface',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'TEST_2': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'TEST_2',
            'lineno': 15,
            'section_name': 'DYNAMIC FUNCTIONS',
            'sections': list([
              'DYNAMIC FUNCTIONS',
            ]),
            'tags': list([
              'my_tag',
              'test',
              'test2',
            ]),
            'title': 'TEST_2',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'TEST_3': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'TEST_3',
            'lineno': 19,
            'section_name': 'DYNAMIC FUNCTIONS',
            'sections': list([
              'DYNAMIC FUNCTIONS',
            ]),
            'test_func': 'Test output of dynamic function; need: TEST_3; args: (); kwargs: {}',
            'title': 'TEST_3',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'TEST_4': dict({
            'content': 'Test dynamic func in tags: [[copy("tags")]]',
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'TEST_4',
            'lineno': 23,
            'section_name': 'DYNAMIC FUNCTIONS',
            'sections': list([
              'DYNAMIC FUNCTIONS',
            ]),
            'tags': list([
              'test_4a',
              'test_4b',
              'TEST_4',
            ]),
            'title': 'TEST_4',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'TEST_5': dict({
            'content': '''
              Test a `link <http://www.[[copy('id')]]>`_
              
              .. spec:: TEST_6
                  :id: TEST_6
              
                  nested id [[copy('id')]]
              
                  nested id also :need_func:`[[copy("id")]]`
              
                  nested id best :ndf:`copy("id")`
            ''',
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'TEST_5',
            'lineno': 29,
            'parent_needs_back': list([
              'TEST_6',
            ]),
            'section_name': 'DYNAMIC FUNCTIONS',
            'sections': list([
              'DYNAMIC FUNCTIONS',
            ]),
            'tags': list([
              'TEST_5',
            ]),
            'title': 'TEST_5',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'TEST_6': dict({
            'content': '''
              nested id [[copy('id')]]
              
              nested id also :need_func:`[[copy("id")]]`
              
              nested id best :ndf:`copy("id")`
            ''',
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'TEST_6',
            'lineno': 35,
            'parent_need': 'TEST_5',
            'parent_needs': list([
              'TEST_5',
            ]),
            'section_name': 'DYNAMIC FUNCTIONS',
            'sections': list([
              'DYNAMIC FUNCTIONS',
            ]),
            'title': 'TEST_6',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'TEST_7': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'TEST_7',
            'lineno': 46,
            'section_name': 'DYNAMIC FUNCTIONS',
            'sections': list([
              'DYNAMIC FUNCTIONS',
            ]),
            'status': 'draft',
            'test_func': "Test output of dynamic function; need: TEST_7; args: ('TEST_7',); kwargs: {'status': 'draft'}",
            'title': 'TEST_7',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'TEST_8': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'TEST_8',
            'lineno': 52,
            'section_name': 'DYNAMIC FUNCTIONS',
            'sections': list([
              'DYNAMIC FUNCTIONS',
            ]),
            'test_func': '??',
            'title': 'TEST_8',
            'type': 'spec',
            'type_name': 'Specification',
          }),
        }),
        'needs_amount': 8,
        'needs_defaults_removed': True,
        'needs_schema': dict({
          '$schema': 'http://json-schema.org/draft-07/schema#',
          'properties': dict({
            'arch': dict({
              'additionalProperties': dict({
                'type': 'string',
              }),
              'default': dict({
              }),
              'description': 'Mapping of uml key to uml content.',
              'field_type': 'core',
              'type': 'object',
            }),
            'avatar': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'closed_at': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'completion': dict({
              'default': '',
              'description': 'Added for needgantt functionality',
              'field_type': 'extra',
              'type': 'string',
            }),
            'constraints': dict({
              'default': list([
              ]),
              'description': 'List of constraint names, which are defined for this need.',
              'field_type': 'core',
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'constraints_error': dict({
              'default': None,
              'description': 'An error message set if any constraint failed, and `error_message` field is set in config.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'constraints_passed': dict({
              'default': True,
              'description': 'True if all constraints passed, False if any failed, None if not yet checked.',
              'field_type': 'core',
              'type': list([
                'boolean',
                'null',
              ]),
            }),
            'constraints_results': dict({
              'additionalProperties': dict({
                'type': 'object',
              }),
              'default': dict({
              }),
              'description': 'Mapping of constraint name, to check name, to result, None if not yet checked.',
              'field_type': 'core',
              'type': list([
                'object',
                'null',
              ]),
            }),
            'content': dict({
              'default': '',
              'description': 'The main content of the need.',
              'field_type': 'core',
              'type': 'string',
            }),
            'created_at': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'docname': dict({
              'default': None,
              'description': 'Name of the document where the need is defined (None if external).',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'doctype': dict({
              'default': '.rst',
              'description': "The markup type of the content, denoted by the suffix of the source file, e.g. '.rst'.",
              'field_type': 'core',
              'type': 'string',
            }),
            'duration': dict({
              'default': '',
              'description': 'Added for needgantt functionality',
              'field_type': 'extra',
              'type': 'string',
            }),
            'external_css': dict({
              'default': '',
              'description': 'CSS class name, added to the external reference.',
              'field_type': 'core',
              'type': 'string',
            }),
            'external_url': dict({
              'default': None,
              'description': 'URL of the need, if it is an external need.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'has_dead_links': dict({
              'default': False,
              'description': 'True if any links reference need ids that are not found in the need list.',
              'field_type': 'core',
              'type': 'boolean',
            }),
            'has_forbidden_dead_links': dict({
              'default': False,
              'description': 'True if any links reference need ids that are not found in the need list, and the link type does not allow dead links.',
              'field_type': 'core',
              'type': 'boolean',
            }),
            'id': dict({
              'description': 'ID of the data.',
              'field_type': 'core',
              'type': 'string',
            }),
            'id_prefix': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'is_external': dict({
              'default': False,
              'description': 'If true, no node is created and need is referencing external url.',
              'field_type': 'core',
              'type': 'boolean',
            }),
            'is_import': dict({
              'default': False,
              'description': 'If true, the need was derived from an import.',
              'field_type': 'core',
              'type': 'boolean',
            }),
            'is_modified': dict({
              'default': False,
              'description': 'Whether the need was modified by needextend.',
              'field_type': 'core',
              'type': 'boolean',
            }),
            'jinja_content': dict({
              'default': False,
              'description': 'Whether the content was pre-processed by jinja.',
              'field_type': 'core',
              'type': 'boolean',
            }),
            'layout': dict({
              'default': None,
              'description': 'Key of the layout, which is used to render the need.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'lineno': dict({
              'default': None,
              'description': 'Line number where the need is defined (None if external).',
              'field_type': 'core',
              'type': list([
                'integer',
                'null',
              ]),
            }),
            'links': dict({
              'default': list([
              ]),
              'description': 'Link field',
              'field_type': 'links',
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'links_back': dict({
              'default': list([
              ]),
              'description': 'Backlink field',
              'field_type': 'backlinks',
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'max_amount': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'max_content_lines': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'modifications': dict({
              'default': 0,
              'description': 'Number of modifications by needextend.',
              'field_type': 'core',
              'type': 'integer',
            }),
            'params': dict({
              'default': '',
              'description': 'Added by service open-needs',
              'field_type': 'extra',
              'type': 'string',
            }),
            'parent_need': dict({
              'default': None,
              'description': 'Simply the first parent id.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'parent_needs': dict({
              'default': list([
              ]),
              'description': 'Link field',
              'field_type': 'links',
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'parent_needs_back': dict({
              'default': list([
              ]),
              'description': 'Backlink field',
              'field_type': 'backlinks',
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'parts': dict({
              'additionalProperties': dict({
                'type': 'object',
              }),
              'default': dict({
              }),
              'description': "Mapping of parts, a.k.a. sub-needs, IDs to data that overrides the need's data",
              'field_type': 'core',
              'type': 'object',
            }),
            'post_content': dict({
              'default': None,
              'description': 'Additional content after the need.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'post_template': dict({
              'default': None,
              'description': 'The template key, if the post_content was created from a jinja template.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'pre_content': dict({
              'default': None,
              'description': 'Additional content before the need.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'pre_template': dict({
              'default': None,
              'description': 'The template key, if the pre_content was created from a jinja template.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'prefix': dict({
              'default': '',
              'description': 'Added by service open-needs',
              'field_type': 'extra',
              'type': 'string',
            }),
            'query': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'section_name': dict({
              'default': None,
              'description': 'Simply the first section.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'sections': dict({
              'default': list([
              ]),
              'description': 'Sections of the need.',
              'field_type': 'core',
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'service': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'signature': dict({
              'default': None,
              'description': 'Derived from a docutils desc_name node.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'specific': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'status': dict({
              'default': None,
              'description': 'Status of the need.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'style': dict({
              'default': None,
              'description': 'Comma-separated list of CSS classes (all appended by `needs_style_`).',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'tags': dict({
              'default': list([
              ]),
              'description': 'List of tags.',
              'field_type': 'core',
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'template': dict({
              'default': None,
              'description': 'The template key, if the content was created from a jinja template.',
              'field_type': 'core',
              'type': list([
                'string',
                'null',
              ]),
            }),
            'test_func': dict({
              'default': '',
              'description': 'Added by needs_extra_options config',
              'field_type': 'extra',
              'type': 'string',
            }),
            'title': dict({
              'description': 'Title of the need.',
              'field_type': 'core',
              'type': 'string',
            }),
            'type': dict({
              'default': '',
              'description': 'Type of the need.',
              'field_type': 'core',
              'type': 'string',
            }),
            'type_name': dict({
              'default': '',
              'description': 'Name of the type.',
              'field_type': 'core',
              'type': 'string',
            }),
            'updated_at': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'url': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
            'url_postfix': dict({
              'default': '',
              'description': 'Added by service open-needs',
              'field_type': 'extra',
              'type': 'string',
            }),
            'user': dict({
              'default': '',
              'description': 'Added by service github-issues',
              'field_type': 'extra',
              'type': 'string',
            }),
          }),
          'type': 'object',
        }),
      }),
    }),
  })
# ---
